### PR TITLE
executor: optimize point lock read process for rc

### LINKED ISSUE #36833: executor: Lock the key only if it exists in the tikv for RC isolation
## Enhancement
1. current process
For RC isolation, TiDB currently  sends a get request to judge whether the keys exists on the TiKV firstly, then sends a pessimistic lock request to TiKV if the key exists, It makes and additional rpc request.

2. what changed
We add a flag LockIfExists for pessimistic lock request and TiKV will check the flag to decide how to execute the lock process. TiDB only send a pessimistic lock request to TiKV. If the keys exist, TiKV locks the keys and return the values of the keys, else does nothing and return.

### PR DESCRIPTION
**Pull request by TonsnakeLin**
Signed-off-by: TonsnakeLin <lpbgytong@163.com>

### What problem does this PR solve?

Issue Number: close #36833

Problem Summary:
or RC isolation, TiDB currently sends a get request to judge whether the keys exists on the TiKV firstly, then sends a pessimistic lock request to TiKV if the key exists, It makes and additional rpc request.

### What is changed and how it works?
We add a flag LockIfExists for pessimistic lock request and TiKV will check the flag to decide how to execute the lock process. TiDB only send a pessimistic lock request to TiKV. If the keys exist, TiKV locks the keys and return the values of the keys, else does nothing and return.

This PR is depended on these PRs bellow:

- [x] https://github.com/tikv/client-go/pull/561

- [x] https://github.com/tikv/tikv/pull/13211

- [x] https://github.com/pingcap/kvproto/pull/963

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by ekexium** [comment]
Does `lockKeyIfNeeded` always return values? If it's followed by a `get`, would it be better to let the lock request return values as well so the `get` can read from cache?

File: executor/point_get.go [hunk 1]
```diff
@@ -354,16 +358,19 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
 func (e *PointGetExecutor) getAndLock(ctx context.Context, key kv.Key) (val []byte, err error) {
 	if e.ctx.GetSessionVars().IsPessimisticReadConsistency() {
 		// Only Lock the exist keys in RC isolation.
```
Comment (ekexium) [inline executor/point_get.go:360]:
```suggestion
		// Only Lock the existing keys in RC isolation.
```

File: executor/point_get.go [hunk 2]
```diff
@@ -354,16 +358,19 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
...
-		val, err = e.get(ctx, key)
-		if err != nil {
-			if !kv.ErrNotExist.Equal(err) {
+		if e.lock {
```
Comment (ekexium) [inline executor/point_get.go:361]: The original function doesn't check `e.lock`. What could be wrong if just do `lockKeyIfExists`?

**Review by ekexium — COMMENTED** [review]
The original implementation reads using `start_ts` while `lockIfExists` seems to read using `for_update_ts`. Do I understand it correctly? Can we prove they are equivalent in this case?

**Comment by TonsnakeLin** [comment]
> Does `lockKeyIfNeeded` always return values? If it's followed by a `get`, would it be better to let the lock request return values as well so the `get` can read from cache?
Yes, lockKeyIfNeeded` always returns value.
Actually, even you call function `lockKeyIfNeeded` first, and then call funciton `get` , it will get  the value from the pessimitic lock cache.
I don't change the RR process of point-get, it locks first, then get the value from cache. So this PR only affect the RC process.
The old process of RC point-get is that it get the value first and lock the key if it exists. Now, I merged the two step together.

File: executor/point_get.go (hunk 2 above)
Comment (TonsnakeLin) [reply executor/point_get.go:361]: In the past, checking `e.lock` is in the function `lockKeyIfNeeded`. If e.lock is false, lockKeyIfNeeded does nothing and returns directly.
I keep the process as old that it gets the value in the function `getAndLock`.

**Comment by TonsnakeLin** [comment]
> The original implementation reads using `start_ts` while `lockIfExists` seems to read using `for_update_ts`. Do I understand it correctly? Can we prove they are equivalent in this case?

No, It always uses for-update-ts to read data at RC isolation level.
here is the stack of point-read statement getting snapshot .
image

**Comment by ekexium** [comment]
> > The original implementation reads using `start_ts` while `lockIfExists` seems to read using `for_update_ts`. Do I understand it correctly? Can we prove they are equivalent in this case?
> 
> No, It always uses for-update-ts to read data at RC isolation level. here is the stack of point-read statement getting snapshot . image

Oh right. I forgot it's RC

**Comment by TonsnakeLin** [comment]
@cfzjywxk @MyonKeminta @sticnarf PTAL, thanks!

File: tests/realtikvtest/pessimistictest/pessimistic_test.go [hunk 3]
```diff
@@ -3381,3 +3384,162 @@ func TestLazyUniquenessCheckWithStatementRetry(t *testing.T) {
 	require.False(t, tk.Session().GetSessionVars().InTxn())
 	tk.MustExec("admin check table t5")
 }
+
+func TestRCPointWriteLockIfExists(t *testing.T) {
```
Comment (MyonKeminta) [inline tests/realtikvtest/pessimistictest/pessimistic_test.go:3388]: Consider covering the cases that the key is already locked in a previous statement of the transaction, including both existing or not-existing cases.

Comment (TonsnakeLin) [reply tests/realtikvtest/pessimistictest/pessimistic_test.go:3388]: I have added the case that  key is already locked in a previous statement when the key exists.
I'll try to add the case that key is already locked in a previous statement but the key doesn't exists.

File: executor/point_get.go [hunk 4]
```diff
@@ -383,28 +390,68 @@ func (e *PointGetExecutor) getAndLock(ctx context.Context, key kv.Key) (val []by
...
+		if LockOnlyIfExists {
+			return e.getValueFromLockCtx(ctx, lockCtx, key)
+		}
 	}
-	return nil
+
+	return nil, nil
+}
+
+func (e *PointGetExecutor) getValueFromLockCtx(ctx context.Context,
```
Comment (ekexium) [inline executor/point_get.go:436]: I think it might be better to put the function of extracting values from LockCtx in client-go. But this is acceptable.

**Review by ekexium — APPROVED** [review]

**Review by MyonKeminta — APPROVED** [review]

File: executor/point_get.go [hunk 5]
```diff
@@ -254,40 +254,44 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
...
-				// Change the unique index LOCK into PUT record.
-				if e.lock && len(e.handleVal) > 0 {
-					if !e.txn.Valid() {
-						return kv.ErrInvalidTxn
-					}
-					memBuffer := e.txn.GetMemBuffer()
-					err = memBuffer.Set(e.idxKey, e.handleVal)
+			} else {
+				if e.lock {
+					e.handleVal, err = e.lockKeyIfExists(ctx, e.idxKey)
```
[the remark this answers is not in the document]
Comment (TonsnakeLin) [reply executor/point_get.go:265]: Um..., (e *PointGetExecutor) Next is a little messy before, especially processing the secondary index, I refactored some code. I think that processing the cluster index key and secondary key index may be unified. But I don't want to modify  more code here, I want to refactor this in a new PR.

[the remark this answers is not in the document]
Comment (cfzjywxk) [reply executor/point_get.go:265]: Better to record some refactor tasks or issues, it's getting messier by now.

File: executor/point_get.go (hunk 4 above)
Comment (cfzjywxk) [reply executor/point_get.go:436]: +1

File: executor/point_get.go [hunk 6]
```diff
@@ -383,28 +390,68 @@ func (e *PointGetExecutor) getAndLock(ctx context.Context, key kv.Key) (val []by
 }
 
 func (e *PointGetExecutor) lockKeyIfNeeded(ctx context.Context, key []byte) error {
+	_, err := e.lockKeyBase(ctx, key, false)
+	return err
+}
+
+// lockKeyIfExists locks the key if needed, but won't lock the key if it doesn't exis.
```
Comment (cfzjywxk) [inline executor/point_get.go:397]:
```suggestion
// lockKeyIfExists locks the key if needed, but won't lock the key if it doesn't exist.
```

**Comment by cfzjywxk** [comment]
@TonsnakeLin 
There're some comments which could be addressed latter.I'll merge it first.

File: executor/point_get.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (TonsnakeLin) [reply executor/point_get.go:265]: #37853

**Pull request merged**